### PR TITLE
Ruff codechecks

### PR DESCRIPTION
**Pull request by PaulJonasJost**
Most important to check would be 
+ pre-commit-config.yaml
+ pyproject.toml

can flake8 be deleted from tox and .flake8 as well?

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -54,12 +31,19 @@ repos:
...
-- repo: local
+- repo: https://github.com/astral-sh/ruff-pre-commit
+  # Ruff version.
+  rev: v0.1.11
   hooks:
-  - id: style
-    name: Check style
-    description: Check style
-    pass_filenames: false
-    entry: tox -e project,flake8
```
Comment (dweindl) [inline .pre-commit-config.yaml:63]: Did you remove `project` on purpose?

If we aren't using flake8 any longer, I'd also remove the respective tox config from `tox.ini`.

File: pypesto/engine/base.py [hunk 2]
```diff
@@ -24,5 +24,6 @@ def execute(
             List of tasks to execute.
         progress_bar:
             Whether to display a progress bar.
+
```
Comment (dweindl) [inline pypesto/engine/base.py:27]: Any good reason for those trailing empty lines?

File: pyproject.toml [hunk 3]
```diff
@@ -11,10 +11,80 @@ build-backend = "setuptools.build_meta"
...
+"test/*" = [
+    "T201",
+    "S101",
+    "D",
+]
+"pypesto/util.py" = [
+    "D400",
+    "D205",
+    "D107",
+]
```
Comment (dweindl) [inline pyproject.toml:81]: Why different docstring rules in different files?

**Review by dweindl — COMMENTED** [review]
Thanks. Mostly looks good.

File: doc/example/workflow_comparison.ipynb [hunk 4]
```diff
@@ -468,7 +195,7 @@
     "        \"\"\"\n",
     "        Turn array of parameters to dictionary usable for objective call.\n",
     "        \"\"\"\n",
-    "        return {x_id: x_i for x_id, x_i in zip(self.x_ids, x)}\n",
+    "        return {zip(self.x_ids, x)}\n",
```
Comment (dweindl) [inline doc/example/workflow_comparison.ipynb:198]: fixme: dict vs set

File: pyproject.toml [hunk 5]
```diff
@@ -11,10 +11,80 @@ build-backend = "setuptools.build_meta"
...
 line-length = 79
-target-version = ['py38', 'py39']
 skip-string-normalization = true
 
 [tool.isort]
 profile = "black"
 line_length = 79
 multi_line_output = 3
+
+[tool.ruff]
```
Comment (dilpath) [inline pyproject.toml:21]: Remove `tool.isort`, `tool.black` above?

**Review by dilpath — APPROVED** [review]

**Comment by stephanmg** [comment]
> Most important to check would be
> 
> * pre-commit-config.yaml
> * pyproject.toml
> 
> can flake8 be deleted from tox and .flake8 as well?

Looks good to me. Would remove .flake8.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (PaulJonasJost) [reply .pre-commit-config.yaml:63]: As far as I can see in the code, project is just used to initialise the project to be able to run flake8. Not needed for any part of the tests if we remove the flake8 test. Yes, agreed, will remove from tox.ini

File: pyproject.toml (hunk 3 above)
Comment (PaulJonasJost) [reply pyproject.toml:82]: * I do not really care about the documentation within `examples`, in many cases I think it is not supposed to be documented the same the code base (hence ignore `D` there)
* Same goes for testing and Jupyter notebooks
* All other files (both */init and the other ones). are the ones that have in their heading the following string which is considered docstring:
```
Header
=====

Short description of either the module or the file
```

This produces exactly D400 and D205, which is why in only those files it is excluded, in all other files it should be checked.

Comment (dweindl) [reply pyproject.toml:82]: Ok

**Review by dweindl — APPROVED** [review]

**Comment by FFroehlich** [comment]
bit late now, but might make sense to add the PR of this commit to a `.git-blame-ignore-revs`  file (see https://docs.github.com/en/repositories/working-with-files/using-files/viewing-a-file#ignore-commits-in-the-blame-view)

**Pull request merged**